### PR TITLE
feat: contract deployment map caching

### PR DESCRIPTION
**Pull request by PatrickAlphaC**
### What I did
I created a function called `find_deployments` that allows you to access a recently deployed contract of a `ContractContainer`. 

fixes: # https://github.com/ApeWorX/ape/issues/809

### How I did it
I made a `_local_contracts_mapping` object and a `_contracts_mapping_cache` that is written to disc that organizes the contracts cache. 

### How to verify it

The following code should 
```python
from ape import project, chain, accounts

def main():
  account = accounts.test_accounts[0]
  account.deploy(project.MyContract)  
  my_contract = chain.contracts.find_deployments(project.MyContract)[-1]  # returns a ContractInstance
```

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by PatrickAlphaC** [comment]
I didn't test the writing to disc of this... but I think this is approximately what we are looking for... would love some help on:

1. Cleaning up the code
2. Testing the writing to disc part, and reading from disc

**Comment by PatrickAlphaC** [comment]
I've manually confirmed this works with live networks as well... but I'd love to do an integration test on it... would love some help there.

File: src/ape/managers/chain.py [hunk 1]
```diff
@@ -403,6 +407,11 @@ def _contract_types_cache(self) -> Path:
         network_name = self._network.name.replace("-fork", "")
         return self._network.ecosystem.data_folder / network_name / "contract_types"
 
+    # I don't love this. I think I'd rather it be at the config level.
+    @property
+    def _contracts_mapping_cache(self) -> Path:
+        return self._network.ecosystem.data_folder / "contracts_map.json"
```
Comment (unparalleled-js) [inline src/ape/managers/chain.py:413]:
```suggestion
        return self._network.ecosystem.data_folder / "deployments.json"
```

I am tryin to think of way to lessen the confusion between this feature and the contracts caching we do.

**Review by unparalleled-js — COMMENTED** [review]
Amazing!

Some work may need to be done involving deploying proxies. I am okay with saving for later though as there is a bug regarding local proxies atm: https://github.com/ApeWorX/ape/pull/837

- [ ] Update `contracts.md` to include some mentionings of this feature

File: src/ape/managers/chain.py [hunk 2]
```diff
@@ -420,16 +432,34 @@ def __setitem__(self, address: AddressType, contract_type: ContractType):
             contract_type (ContractType): The contract's type.
         """
 
-        if self.get(address) and self._is_live_network:
-            return
+        # Do we need this?
+        # if self.get(address) and self._is_live_network:
```
Comment (PatrickAlphaC) [inline src/ape/managers/chain.py:436]: I have no idea why we need this line... If we need it, I'll have to make some sort of "don't make the proxy call" or "don't check the network" flag for testing. Let me know what you think.

**Comment by PatrickAlphaC** [comment]
```python
pytest tests/functional/test_chain.py
```

This is passing all the tests in `test_chain.py`. I can't seem to get `pytest` to pass all the tests even on the main branch (like, `git clone ape`, `cd ape`, `pytest` isn't working for me after setting up a venv... maybe I'm doing something wrong.)

**Comment by PatrickAlphaC** [comment]
Added changes, you can now do this:

```
def main():
  account = accounts.test_accounts[0]
  account.deploy(project.MyContract)  
  my_contract = project.MyContract.deployments[-1]  # returns a ContractInstance
```

Great suggestion @unparalleled-js

File: docs/userguides/contracts.md [hunk 3]
```diff
@@ -36,6 +36,32 @@ from ape import Contract
 contract = Contract("v2.registry.ychad.eth")
 ```
 
+## From Previous Deployment
+
+Ape keeps track of your deployments for you, so you can always refer back to a version that you deployed previously. On live networks, this history of deployments is saved, on local networks, this history lasts for the duration of your script.
```
Comment (unparalleled-js) [inline docs/userguides/contracts.md:41]: A general pattern we have been following with markdown is "1 sentence per line"

File: src/ape/contracts/base.py [hunk 4]
```diff
@@ -667,6 +667,10 @@ def __init__(self, contract_type: ContractType) -> None:
     def __repr__(self) -> str:
         return f"<{self.contract_type.name}>"
 
+    @property
+    def deployments(self):
```
Comment (unparalleled-js) [inline src/ape/contracts/base.py:671]: - [ ] property method doc

File: src/ape/managers/chain.py [hunk 5]
```diff
@@ -403,6 +410,11 @@ def _contract_types_cache(self) -> Path:
         network_name = self._network.name.replace("-fork", "")
         return self._network.ecosystem.data_folder / network_name / "contract_types"
 
+    # I don't love this. I think I'd rather it be at the config level.
```
Comment (unparalleled-js) [inline src/ape/managers/chain.py:413]: will others part of the code need it? If no, we can leave it for now.
keeping it underscored will allow us to change it later if needbe

File: src/ape/managers/chain.py [hunk 6]
```diff
@@ -553,6 +584,46 @@ def instance_at(
 
         return Address(address)
 
+    def find_deployments(self, contract_container: ContractContainer) -> List[ContractInstance]:
+        """Finds previous deployments of a contract container or contract type.
+        Locally deployed contracts are saved for the duration of the script and read from
+        _local_deployments_mapping, while those deployed on a live network are written to disc
```
Comment (unparalleled-js) [inline src/ape/managers/chain.py:590]:
```suggestion
        ``_local_deployments_mapping``, while those deployed on a live network are written to disk
```

File: src/ape/managers/chain.py [hunk 7]
```diff
@@ -553,6 +584,46 @@ def instance_at(
...
+        in deployments_map.json.
+
+        Args:
+            contract_container (ContractContainer): The ContractContainer you want to get a list of previous deployments for.
```
Comment (unparalleled-js) [inline src/ape/managers/chain.py:594]:
```suggestion
            contract_container (:class:`~ape.contracts.ContractContainer`): The ContractContainer you want to get a list of previous deployments for.
```

**Review by unparalleled-js — COMMENTED** [review]
Hey this looks good!

Are you still wondering if certain changes or needed? That if statement was just an optimization, not logically necessary.

If you are confident that things are working and these tests pass, let's clean up the questions / commented out code.  I am going to do some QA right now ina minute

**Comment by unparalleled-js** [comment]
Found a pretty serious bug while testing this issue!

When you deploy a new contract, it removes all deployments from every other contract

- [ ] fix + add a test

File: tests/functional/test_contract_container.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    contract_from_cache = Contract(contract.address)
+    assert contract_from_cache.contract_type == contract.contract_type
+    assert contract_from_cache.address == contract.address
+
+    # Clean up for next test
+    del chain.contracts._local_contracts[contract_from_cache.address]
+
+
+def test_deployment_property(chain, owner, project_with_contract, eth_tester_provider):
+    initial_deployed_contract = owner.deploy(project_with_contract.ApeContract)
```
Comment (unparalleled-js) [inline tests/functional/test_contract_container.py:22]:
```suggestion
    initial_deployed_contract = owner.deploy(project_with_contract.ApeContract0)
```

File: tests/functional/test_contract_container.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    assert (
+        project_with_contract.ApeContract.deployments[-1].address
```
Comment (unparalleled-js) [inline tests/functional/test_contract_container.py:24]:
```suggestion
        project_with_contract.ApeContract0.deployments[-1].address
```

File: tests/functional/test_chain.py [hunk 10]
```diff
@@ -219,3 +227,136 @@ def test_instance_at_when_given_name_as_contract_type(chain, contract_instance):
...
+    # Arrange / Act
+    initial_deployed_contract = owner.deploy(project_with_contract.ApeContract0)
+    owner.deploy(project_with_contract.ApeContract0)
+    owner.deploy(project_with_contract.ApeContract0)
+    final_deployed_contract = owner.deploy(project_with_contract.ApeContract0)
+
+    my_contracts_list = chain.contracts.get_deployments(project_with_contract.ApeContract0)
+
+    # Assert
+    assert len(my_contracts_list) == 4
```
Comment (unparalleled-js) [inline tests/functional/test_chain.py:360]: Either need to assert that this is 4 + more than before running the test (like grab the start count in the Arrange section)..

OR

use the fixture that clears the cache before + after.

**Review by unparalleled-js — APPROVED** [review]
LGTM

**Pull request merged**